core/txpool/bundlepool/bundlepool.go

```diff
@@ -0,0 +1,439 @@
+func (p *BundlePool) Content() (map[common.Address][]*types.Transaction, map[common.Address][]*types.Transaction) {
+	// TODO implement me
+	panic("implement me")
+}
+
+// ContentFrom retrieves the data content of the transaction pool, returning the
+// pending as well as queued transactions of this address, grouped by nonce.
+func (p *BundlePool) ContentFrom(addr common.Address) ([]*types.Transaction, []*types.Transaction) {
+	// TODO implement me
+	panic("implement me")
+}
+
+// Locals retrieves the accounts currently considered local by the pool.
+func (p *BundlePool) Locals() []common.Address {
+	// TODO implement me
+	panic("implement me")
+}
+
+// Status returns the known status (unknown/pending/queued) of a transaction
+// identified by their hashes.
+func (p *BundlePool) Status(hash common.Hash) txpool.TxStatus {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (p *BundlePool) filter(tx *types.Transaction) bool {
+	switch tx.Type() {
+	case types.LegacyTxType, types.AccessListTxType, types.DynamicFeeTxType:
+		return true
+	default:
+		return false
+	}
+}
+
+func (p *BundlePool) reset(oldHead, newHead *types.Header) {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (p *BundlePool) addBundles(bundles []*types.Bundle) []error {
+	errs := make([]error, len(bundles))
+	for i, bundle := range bundles {
+		if err := p.AddBundle(bundle); err != nil {
+			errs[i] = err
+		}
+	}
+	return errs
+}
+
+func (p *BundlePool) drop() {
+	leastPrice := big.NewInt(math.MaxInt64)
+	leastBundleHash := common.Hash{}
+	for h, b := range p.bundles {
+		if b.Price.Cmp(leastPrice) < 0 {
+			leastPrice = b.Price
+			leastBundleHash = h
+		}
+	}
+	p.PruneBundle(leastBundleHash)
+}
+
+// =====================================================================================================================
+
+// NewBundleGasPricer creates a new BundleGasPricer.
+func NewBundleGasPricer(expire time.Duration) *BundleGasPricer {
+	return &BundleGasPricer{
+		expire: expire,
+		queue:  prque.New[int64, *gasPriceInfo](nil),
+		latest: common.Big0,
+	}
+}
+
+// BundleGasPricer is a limited number of queues.
+// In order to avoid too drastic gas price changes, the latest n gas prices are cached.
+// Allowed as long as the user's Gas Price matches this range.
+type BundleGasPricer struct {
+	mu     sync.RWMutex
+	expire time.Duration
+	queue  *prque.Prque[int64, *gasPriceInfo]
+	latest *big.Int
+}
+
+type gasPriceInfo struct {
+	val  *big.Int
+	time time.Time
+}
+
+// Push is a method to cache a new gas price.
+func (pool *BundleGasPricer) Push(gasPrice *big.Int) {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+	pool.retire()
+	index := -gasPrice.Int64()
+	pool.queue.Push(&gasPriceInfo{val: gasPrice, time: time.Now()}, index)
+	pool.latest = gasPrice
+}
+
+func (pool *BundleGasPricer) retire() {
+	now := time.Now()
+	for !pool.queue.Empty() {
+		v, _ := pool.queue.Peek()
+		info := v
+		if info.time.Add(pool.expire).After(now) {
+			break
```

Comment:
Hardly to understand this, the queue is sorted by queue not time, how can you ensure no other item is also expired but break here?

Reply:
We only need to make sure that the minimal price is not expired

Reply:
It is possible OOM here, lets us say the expire time is 5 minutes, every minutes there is a bundle with gas price to be 1wei less than the last one, the `BundleGasPricer` can never expire one item.

Reply:
Fixed. By the way, these codes are copied from bsc-private. The code over there also has this problem.

Reply:
Every time you push gasprice into `BundleGasPricer`, you will do `retire`, and it will make it only keep one item in `BundleGasPricer`, so what is the meaning to keep a `BundleGasPricer`?

Did not find similar code in bsc-private repo, would you share a link?

Reply:
https://github.com/bnb-chain/bsc-private/blob/2f44f164a4d73091e59108f55af0ecc2c3c9783b/core/mev_gas_price_pool.go#L56